django/db/models/expressions.py

```diff
@@ -380,77 +385,87 @@ def as_sql(self, compiler, connection):
+            dispatchers = (self.expression_dispatchers.get(self.connector),
+                           self.expression_dispatchers.get(CombinedExpression.ANY))
 
+            for dispatcher in filter(bool, dispatchers):
+                try:
+                    klass = dispatcher(lhs_output or fields.Field(), rhs_output or fields.Field())
+                except NotImplementedError:
+                    continue
 
-class DurationExpression(CombinedExpression):
-    def compile(self, side, compiler, connection):
-        if not isinstance(side, DurationValue):
-            try:
-                output = side.output_field
-            except FieldError:
-                pass
-            else:
-                if output.get_internal_type() == 'DurationField':
-                    sql, params = compiler.compile(side)
-                    return connection.ops.format_for_duration_arithmetic(sql), params
-        return compiler.compile(side)
+                # avoid recursion in case CombinedExpression maps to another
+                # CombinedExpression, but with a different connector
+                if type(self) != klass:
+                    try:
+                        output_field = self._output_field_or_none
+                    except FieldError:
+                        output_field = None
+
+                    expression = klass(self.lhs, self.connector, self.rhs, output_field=output_field)
+                    return compiler.compile(expression)
+
+        return self._as_sql(compiler, connection)
+
+    def _as_sql(self, compiler, connection, compile=None, combine_expression=None):
+        compile = compile or compiler.compile
+        combine_expression = combine_expression or connection.ops.combine_expression
 
-    def as_sql(self, compiler, connection):
-        connection.ops.check_expression_support(self)
         expressions = []
         expression_params = []
-        sql, params = self.compile(self.lhs, compiler, connection)
+        sql, params = compile(self.lhs)
         expressions.append(sql)
         expression_params.extend(params)
-        sql, params = self.compile(self.rhs, compiler, connection)
+        sql, params = compile(self.rhs)
         expressions.append(sql)
         expression_params.extend(params)
         # order of precedence
         expression_wrapper = '(%s)'
-        sql = connection.ops.combine_duration_expression(self.connector, expressions)
+        sql = combine_expression(self.connector, expressions)
         return expression_wrapper % sql, expression_params
 
+    @classmethod
+    def dispatch(cls, connector, lhs, rhs, commutative=True):
+        dispatcher = cls.expression_dispatchers.setdefault(connector, multipledispatch.Dispatcher(connector))
 
-class TemporalSubtraction(CombinedExpression):
-    def __init__(self, lhs, rhs):
-        super(TemporalSubtraction, self).__init__(lhs, self.SUB, rhs, output_field=fields.DurationField())
+        def _inner(klass):
+            try:
```

Comment:
Is this try/except clause as minimal as can be? It's not obvious where the `TypeError` might come from. Some comments may help.